resteasy-cdi/src/main/java/org/jboss/resteasy/cdi/ResteasyCdiExtension.java

```diff
@@ -81,9 +87,27 @@ public static boolean isCDIActive()
    public void observeBeforeBeanDiscovery(@Observes BeforeBeanDiscovery event, BeanManager beanManager)
    {
       this.beanManager = beanManager;
+      final AnnotatedType<ContextProducers> producersAnnotatedType = beanManager.createAnnotatedType(ContextProducers.class);
+      event.addAnnotatedType(producersAnnotatedType, ContextProducers.class.getCanonicalName());
       active = true;
    }
 
+
+   /**
+    * Registers a producer and disable for a {@link Client REST client}.
+    *
+    * @param event       the after bean discovery event
+    * @param beanManager the bean manager
+    */
+   public void registerClientProducer(@Observes AfterBeanDiscovery event, final BeanManager beanManager) {
```

Comment:
You should probably add a test that verifies this registration doesn't happen if the bean is user-provided.

Otherwise, to answer your email question @jamezp - the idea of conditional bean addition is fine however I wouldn't recommend using `beanManager.getBeans()` to determine whether there are such beans.

The reason is that BM is caller sensitive and in EE implements some visibility limitations. Therefore, `getBeans()` method returns whichever beans that given BM "can see" so to speak.
A more universal approach would be to add an extension method observing `ProcessBean<X>`. When there is a bean of type `X`, this observer will be notified and you can inspect the bean and store the information that you no longer need to provide a synthetic one. Here's the code:

```
private boolean generateClientBean = true;

public void observeProcessBean(@Observes ProcessBean<Client> pb) {
    // this method will get notified if there is any bean of type `Client` created
    Set<Annotation> qualifiers = pb.getBean().getQualifiers(); // you want to detect beans with no explicit qualifiers
    if (qualifiers.contains(Any.Literal.INSTANCE) && qualifiers.contains(Any.Literal.INSTANCE)) {
        generateClientBean = false;
    }
}

public void registerClientProducer(@Observes AfterBeanDiscovery event) {
    if (generateClientBean) {
        event.addBean().addTransitiveTypeClosure(Client.class)
                 .scope(ApplicationScoped.class)
                 .produceWith(instance -> ClientBuilder.newClient(RegisterBuiltin.getClientInitializedResteasyProviderFactory(getClassLoader())))
                 .disposeWith((client, instance) -> client.close());
    }
}
```

Note that calling `AfterBeanDiscovery.addBean()` will also fire a `ProcessSyntheticBean` event (subtype of `ProcessBean`) as [defined by specification](https://jakarta.ee/specifications/cdi/2.0/cdi-spec-2.0.html#after_bean_discovery). But that should do no harm in this case.

You could use similar approach for whichever type of bean you need to add conditionally.
Ask away if there is anything unclear.

Reply:
@manovotn Excellent thanks! This makes sense to me and it didn't feel right to use the `BeanManager` there. I just wasn't sure what the best solution would be :)

I did notice though that using:
```java
public void processClientBean(@Observes final ProcessBean<Client> pb) {
   // this method will get notified if there is any bean of type `Client` created
   final Set<Annotation> qualifiers = pb.getBean().getQualifiers(); // you want to detect beans with no explicit qualifiers
   if (qualifiers.contains(Any.Literal.INSTANCE) && qualifiers.contains(Default.Literal.INSTANCE)) {
      generateClientBean = false;
   }
}
```
Didn't get triggered. It did work with something like:
```java
public <T> void processClientBean(@Observes final ProcessProducer<T, Client> pb) {
   generateClientBean = false;
}
```
I'm not sure if that is correct though. FWIW I defined my producer in my REST resource like:
```java

    @ApplicationScoped
    @jakarta.enterprise.inject.Produces
    public Client createClient() {
        return ClientBuilder.newClient();
    }
```
It seems odd to have it there for sure though so maybe that is my problem.